source/common/upstream/health_checker_impl.cc

```diff
@@ -405,6 +405,22 @@ void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onEvent(uint32_t events)
       events & Network::ConnectionEvent::LocalClose) {
     parent_.dispatcher_.deferredDelete(std::move(client_));
   }
+
+  if ((events & Network::ConnectionEvent::Connected) && parent_.receive_bytes_.empty()) {
```

Comment:
Per your description above, should we also check that send_bytes is empty here? (Not sure if we do, and/or want to support "one-way" modes, relying on transmission acks.)